core/domain/classifier_services_test.py

```diff
@@ -0,0 +1,208 @@
+        self.assertEquals(
+            len(self.string_classifier._l_dp),
+            self.string_classifier._num_docs)
+        self.assertEquals(
+            len(self.string_classifier._c_dl),
+            self.string_classifier._num_docs)
+        if self.string_classifier._num_docs > 0:
+            self.assertEquals(
+                len(self.string_classifier._c_dl[0]),
+                self.string_classifier._num_labels)
+        self.assertEquals(
+            len(self.string_classifier._c_lw),
+            self.string_classifier._num_labels)
+        if self.string_classifier._num_labels > 0:
+            self.assertEquals(
+                len(self.string_classifier._c_lw[0]),
+                self.string_classifier._num_words)
+        self.assertEquals(
+            len(self.string_classifier._c_l),
+            self.string_classifier._num_labels)
+
+    def test_valid_state(self):
+        self.assertEquals(self.string_classifier._num_labels, 3)
+        self.assertEquals(self.string_classifier._num_docs, 2)
+        self.assertEquals(self.string_classifier._num_words, 7)
+        self._validate_instance(self.string_classifier)
+
+    def test_add_train_examples(self):
+        self.string_classifier.add_examples_for_training(
+            self._NEW_EXAMPLES_TRAIN)
+        self.assertEquals(self.string_classifier._num_labels, 3)
+        self.assertEquals(self.string_classifier._num_docs, 3)
+        self.assertEquals(self.string_classifier._num_words, 10)
+        self._validate_instance(self.string_classifier)
+
+    def test_add_test_examples(self):
+        self.string_classifier.add_examples_for_predicting(self._EXAMPLES_TEST)
+        self.assertEquals(self.string_classifier._num_labels, 3)
+        self.assertEquals(self.string_classifier._num_docs, 5)
+        self.assertEquals(self.string_classifier._num_words, 34)
+        self._validate_instance(self.string_classifier)
+
+    def test_empty_load(self):
+        self.string_classifier.load_examples([])
+        # Still got the default label
+        self.assertEquals(self.string_classifier._num_labels, 1)
+        self.assertEquals(self.string_classifier._num_docs, 0)
+        self.assertEquals(self.string_classifier._num_words, 0)
+        self._validate_instance(self.string_classifier)
+
+    def test_empty_add(self):
+        self.string_classifier.add_examples_for_training([])
+        self.assertEquals(self.string_classifier._num_labels, 3)
+        self.assertEquals(self.string_classifier._num_docs, 2)
+        self.assertEquals(self.string_classifier._num_words, 7)
+        self._validate_instance(self.string_classifier)
+
+    def test_model_save_load(self):
+        # Also tests deepcopy
+        # i.e. that editing the dictionary won't edit the model, and vice versa
+        self.assertEquals(
+            self.string_classifier._num_docs,
+            len(self._EXAMPLES_TRAIN))
+        model = self.string_classifier.to_dict()
+        model['_num_docs'] = 9
```

Comment:
Did you see my reply to your reply to my comment in the first pass about breaking up the test a bit?